File: week13/23/prob1_last_stone_weight/prob1.py
# Last Stone in Python
class Solution(object):
    def lastStoneWeight(self, stones):
        """
        :type stones: List[int]
        :rtype: int
        """
        stones.sort()
        while stones:
            x = stones.pop()
            if not stones:
                return x
            else:
                y = stones.pop()
                if x != y:
                    stones.append(x-y)
                    stones.sort()
        return 0

stones = [2,7,4,1,8,1]
sol = Solution()
print(sol.lastStoneWeight(stones))

File: week13/5/prob1.cpp
#include <vector>
#include <algorithm>
#include <functional>
using namespace std;

class Solution {
public:
    int lastStoneWeight(vector<int>& stones) {
        // base case 1
        if (stones.size() == 1) {
            return stones[0];
        }
        // base case 2
        sort(stones.begin(), stones.end(), greater<int>());
        if (stones[1] == 0) {
            return stones[0];
        }
        // recursive case
        stones[0] -= stones[1];
        stones[1] = 0;
        return lastStoneWeight(stones);
    }
};
